In [2]:
import csv
import numpy as np
from textblob import TextBlob
import time

In [3]:
def evaluate(truepositive, truenegative, falsepositive, falsenegative, beta=1):
    accuracy = (truenegative + truepositive)/(truenegative + truepositive + falsenegative + falsepositive)
    precision = truepositive/(truepositive+falsepositive)
    recall = truepositive/(truepositive+falsenegative)
    f = (1+beta**2) * (((precision*beta**2)*recall)/(precision+recall))
    

    return [f, accuracy, precision, recall], truepositive, truenegative, falsepositive, falsenegative

In [4]:
start = time.time()
with open('test.csv', encoding='utf-8') as csvdatei:
    csv_reader_object = csv.reader(csvdatei)
    zeile = 0;
    truepositive = 0
    truenegative = 0
    falsepositive = 0
    falsenegative = 0
    
    alle = []
    
    learningcurve = []
    
    for row in csv_reader_object:
        if zeile < 400000:
            rate = 0
            if int(row[0]) == 2:
                rate = 1
            elif int(row[0]) == 1:
                rate = -1
            rating = TextBlob(row[1] + " " +row[2]).sentiment.polarity
            if rating < 0:
                rating = -1
            else:
                rating = 1
            if rate == rating:
                if rate == 1:
                    truepositive+=1
                    alle.append(1)
                elif rate == -1:
                    truenegative += 1
                    alle.append(1)
            else: 
                if rate == 1:
                    falsepositive+=1
                    alle.append(-1)
                elif rate == -1:
                    falsenegative += 1
                    alle.append(-1)
            zeile +=1
            if zeile % 50000 == 0 and zeile != 0:
                print("Zeile: " + str(zeile))
                print(evaluate(truepositive, truenegative, falsepositive, falsenegative))
ende = time.time()
timetrainwithcurve = '{:5.3f}s'.format(ende-start)

ev = evaluate(truepositive, truenegative, falsepositive, falsenegative)

print("Zufällig auswählen ob positiv oder negativ")
print("Accuracy: " + str(ev[0][1]))
print("Precision: " + str(ev[0][2]))
print("Recall: " + str(ev[0][3]))
print("F: " + str(ev[0][0]))

print("True positive: " + str(ev[1]))
print("True negative: " + str(ev[2]))
print("False positive: " + str(ev[3]))
print("False negative: " + str(ev[4]))

Zeile: 50000
([0.7588757627303797, 0.69098, 0.9582249546780168, 0.6281875726650303], 24314, 10235, 1060, 14391)
Zeile: 100000
([0.7570787765870002, 0.68969, 0.9560291820716107, 0.6266685674295638], 48355, 20614, 2224, 28807)
Zeile: 150000
([0.7560127060959009, 0.6881533333333333, 0.9569402630327999, 0.6248200229336046], 72471, 30752, 3261, 43516)
Zeile: 200000
([0.7549883353917936, 0.687555, 0.9573708546711083, 0.6232392542723977], 96278, 41233, 4287, 58202)
Zeile: 250000
([0.7538067400491031, 0.686336, 0.9573132804899444, 0.6216547910021127], 120049, 51535, 5353, 73063)
Zeile: 300000
([0.7531802619593555, 0.68568, 0.9573665333608373, 0.6207807113301088], 143874, 61830, 6407, 87889)
Zeile: 350000
([0.7529722707722408, 0.6854257142857143, 0.9574895435689382, 0.6204465117654889], 167801, 72098, 7450, 102651)
Zeile: 400000
([0.7522612433498621, 0.684745, 0.957275, 0.6195714082301012], 191455, 82443, 8545, 117557)
Zufällig auswählen ob positiv oder negativ
Accuracy: 0.684745
Precision: 0.9

### Ergebnisse:

Accuracy: 0.684745 <br>
Precision: 0.957275<br>
Recall: 0.6195714082301012<br>
F: 0.7522612433498621<br>
True positive: 191455<br>
True negative: 82443<br>
False positive: 8545<br>
False negative: 117557<br>